### PR TITLE
MINOR: docs(config providers): adds a description of configuration providers

### PR DESCRIPTION
**Pull request by PaulRMellor**
**Documentation**
Adds a section to the docs describing configuration providers and the parameters to use with them, plus a simple example to highlight their use.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: docs/toc.html [hunk 1]
```diff
@@ -22,61 +22,62 @@
     <ul class="toc">
         <li><a href="#gettingStarted">1. Getting Started</a>
             <ul>
-                <li><a href="#introduction">1.1 Introduction</a>
-                <li><a href="#uses">1.2 Use Cases</a>
-                <li><a href="#quickstart">1.3 Quick Start</a>
-                <li><a href="#ecosystem">1.4 Ecosystem</a>
-                <li><a href="#upgrade">1.5 Upgrading</a>
-                <li><a href="#docker">1.6 Docker</a>
+                <li><a href="#introduction">1.1 Introduction</a></li>
```
Comment (mimaison) [inline docs/toc.html]: As per the HTML spec, the closing `li` tag is optional: https://www.w3.org/TR/2012/WD-html-markup-20121025/li.html#li-tags

So let's undo these changes.

File: docs/ops.html [hunk 2]
```diff
@@ -1247,7 +1247,168 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
...
+  <h3 class="anchor-heading">
+    <a id="config_providers" class="anchor-link"></a>
+    <a href="#config_providers">6.6 Configuration Providers</a>
+  </h3>
+  
+  <p>
+    Use configuration providers to load configuration data from external sources. This might include sensitive information, such as passwords, API keys, or other credentials.
+  </p>
+  <p>
+    These configurations can be stored securely and dynamically retrieved as needed. Configuration providers are primarily used with Kafka Connect. For example, you can use providers to supply the credentials for Kafka Connect connectors.
```
Comment (mimaison) [inline docs/ops.html]: `Configuration providers are primarily used with Kafka Connect.`: I don't think this is true. Configuration providers are used by all components. I wonder if we could drop this paragraph altogether.

File: docs/ops.html [hunk 3]
```diff
@@ -1247,7 +1247,168 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
...
+  </p>
+  
+  <p>You have the following options:</p>
+  <ul>
+    <li>
+      Use a custom provider by creating a Java class implementing the
+      <code>ConfigProvider</code> interface and packaging it into a JAR file.
```
Comment (mimaison) [inline docs/ops.html]: Could we link to the javadoc for `ConfigProvider`?
To point to the latest javadoc, you can use `{{version}}/javadoc/org/apache/kafka/common/config/provider/ConfigProvider.html`

Comment (mimaison) [reply docs/ops.html]: Same below for the built-in providers

File: docs/ops.html [hunk 4]
```diff
@@ -1247,7 +1247,168 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
...
+config.providers.fileProvider.class=org.apache.kafka.common.config.provider.FileConfigProvider
+config.providers.fileProvider.config.path=/path/to/connector-properties</code></pre>
+  
+  <p>Next, reference the properties from the file in the connector configuration.</p>
+  <p><b>Example connector configuration referencing file properties</b></p>
+  <pre><code>database.user=${file:/path/to/connector.properties:dbUsername}
+database.password=${file:/path/to/connector.properties:dbPassword}</code></pre>
+  <p>The configuration provider reads and extracts the values from the properties file.</p>
+  
+    <h3 class="anchor-heading"><a id="java" class="anchor-link"></a><a href="#java">6.6 Java Version</a></h3>
```
Comment (mimaison) [inline docs/ops.html]: We added a new section at 6.6, so we need to bump all section numbers below

File: docs/ops.html [hunk 5]
```diff
@@ -1247,7 +1247,168 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
...
+    <dt><b><code>FileConfigProvider</code></b></dt>
+    <dd>
+      <p><code>${fileProvider:&lt;path_and_filename&gt;:&lt;property&gt;}</code></p>
+    </dd>
+  </dl>
+  <p>
+    Here’s an example that uses a file configuration provider with Kafka Connect to provide authentication credentials to a database for a connector.
+  </p>
+  <p>
+    First, create a <code>connector.properties</code> configuration file with the following credentials:
```
Comment (mimaison) [inline docs/ops.html]: Should we name that file `credentials.properties`, or `connector-credentials.properties` so it's not confused with the actual connector configurations that can also be provided as a file.

File: docs/ops.html [hunk 6]
```diff
@@ -1247,7 +1247,168 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
...
+  </p>
+  
+  <p><b>Example Kafka Connect configuration with an <code>FileConfigProvider</code></b></p>
+  <pre><code>config.providers=fileProvider
+config.providers.fileProvider.class=org.apache.kafka.common.config.provider.FileConfigProvider
+config.providers.fileProvider.config.path=/path/to/connector-properties</code></pre>
+  
+  <p>Next, reference the properties from the file in the connector configuration.</p>
+  <p><b>Example connector configuration referencing file properties</b></p>
+  <pre><code>database.user=${file:/path/to/connector.properties:dbUsername}
```
Comment (mimaison) [inline docs/ops.html]: In this example, `file` should be `fileProvider` as the first part of the configuration provider magic string is the alias given to the provider.

**Review by mimaison — COMMENTED** [review]
Thanks for the PR! I left a few comments and suggestions.

**Comment by PaulRMellor** [comment]
Thanks for the comments @mimaison

File: docs/ops.html [hunk 7]
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
...
+  </h3>
+  
+  <p>
+    Use configuration providers to load configuration data from external sources. This might include sensitive information, such as passwords, API keys, or other credentials.
+  </p>
+  
+  <p>You have the following options:</p>
+  <ul>
+    <li>
+      Use a custom provider by creating a Java class implementing the
```
Comment (mimaison) [inline docs/ops.html]: I think we can remove `Java`. You could also use Scala for example to implement a custom ConfigProvider

File: docs/ops.html [hunk 8]
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
...
+  <pre><code>config.providers.provider1.param1=value1
+config.providers.provider1.param2=value2</code></pre>
+  
+  <h4 class="anchor-heading">
+    <a id="config_provider" class="anchor-link"></a>
+    <a href="#config_provider">ConfigProvider</a>
+  </h4>
+  
+  <p>
+    The <code>ConfigProvider</code> is a base interface for all configuration providers. Custom implementations of this interface can retrieve configuration data from various sources. You can then package and deploy the implementation as a JAR file and reference the provider's class in your configuration.
```
Comment (mimaison) [inline docs/ops.html]: I think `deploy` is not the right term here. You just need to add the JAR to the classpath.

File: docs/ops.html [hunk 9]
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
...
+    <li>Use a built-in provider:</li>
+    <ul>
+      <li><a href="/{{version}}/javadoc/org/apache/kafka/common/config/provider/DirectoryConfigProvider.html"><code>DirectoryConfigProvider</code></a></li>
+      <li><a href="/{{version}}/javadoc/org/apache/kafka/common/config/provider/EnvVarConfigProvider.html"><code>EnvVarConfigProvider</code></a></li>
+      <li><a href="/{{version}}/javadoc/org/apache/kafka/common/config/provider/FileConfigProvider.html"><code>FileConfigProvider</code></a></li>
+    </ul>
+  </ul>
+  
+  <p>
+    To use a configuration provider, specify it in your configuration file using the <code>config.providers</code> property.
```
Comment (mimaison) [inline docs/ops.html]: We can remove `file` in this sentence. You can also build a configuration programmatically.

File: docs/ops.html [hunk 10]
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
...
+  <p>
+    To use a configuration provider, specify it in your configuration file using the <code>config.providers</code> property. 
+  </p>
+  <p>
+    Each provider can have its own set of parameters, which are passed in a specific format.
+  </p>
+  
+  <h4 class="anchor-heading">
+    <a id="common_provider_params" class="anchor-link"></a>
+    <a href="#common_provider_params">Common Parameters</a>
```
Comment (mimaison) [inline docs/ops.html]: What about renaming this to something like "Using Configuration Providers"? as effectively this describe how to use them.

File: docs/ops.html [hunk 11]
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
...
+      <li><a href="/{{version}}/javadoc/org/apache/kafka/common/config/provider/EnvVarConfigProvider.html"><code>EnvVarConfigProvider</code></a></li>
+      <li><a href="/{{version}}/javadoc/org/apache/kafka/common/config/provider/FileConfigProvider.html"><code>FileConfigProvider</code></a></li>
+    </ul>
+  </ul>
+  
+  <p>
+    To use a configuration provider, specify it in your configuration file using the <code>config.providers</code> property. 
+  </p>
+  <p>
+    Each provider can have its own set of parameters, which are passed in a specific format.
```
Comment (mimaison) [inline docs/ops.html]: If we rename the section below, we could move this paragraph in there.

File: docs/ops.html [hunk 12]
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
...
+  
+  <p>All configuration providers share a common way of passing parameters.</p>
+  <p>
+    You specify a comma-separated list of aliases and the fully-specified class that implements the configuration provider:
+  </p>
+  <pre><code>config.providers=provider1,provider2
+config.providers.provider1.class=com.example.Provider1
+config.providers.provider2.class=com.example.Provider2</code></pre>
+  <p>
+    Pass additional parameters to each provider as <code>config.providers.&lt;provider_alias&gt;.&lt;parameter_key&gt;=&lt;parameter_value&gt;</code>:
```
Comment (mimaison) [inline docs/ops.html]: The syntax is not correct. It should be `config.providers.<alias>.param.<name>=<value>`.

Some of the examples below are missing `.param.` too.

File: docs/ops.html [hunk 13]
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
...
+      <p><code>${fileProvider:&lt;path_and_filename&gt;:&lt;property&gt;}</code></p>
+    </dd>
+  </dl>
+  <p>
+    Here’s an example that uses a file configuration provider with Kafka Connect to provide authentication credentials to a database for a connector.
+  </p>
+  <p>
+    First, create a <code>connector-credentials.properties</code> configuration file with the following credentials:
+  </p>
+  <p><b>Example configuration file</b></p>
```
Comment (mimaison) [inline docs/ops.html]: I think we can remove this paragraph.

File: docs/ops.html [hunk 14]
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
...
+<pre><code>dbUsername=my-username
+dbPassword=my-password</code></pre>
+  <p><em>NOTE: Ensure the appropriate permissions are set for securely accessing the properties file.
+    For additional security, consider encrypting sensitive information using a secret and implementing decryption in your application at runtime.
```
Comment (mimaison) [inline docs/ops.html]: This tip is not really applicable so I wonder if we drop it. If you're using a configuration provider to provide JAAS credentials for a consumer for example, you can't have some runtime decryption logic. The retrieved values from the provider are directly used.

File: docs/ops.html [hunk 15]
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
...
+  </em></p>
+  <p>
+    Specify a <code>FileConfigProvider</code> in the Kafka Connect configuration:
+  </p>
+  
+  <p><b>Example Kafka Connect configuration with an <code>FileConfigProvider</code></b></p>
+  <pre><code>config.providers=fileProvider
+config.providers.fileProvider.class=org.apache.kafka.common.config.provider.FileConfigProvider
+config.providers.fileProvider.config.path=/path/to/connector-properties</code></pre>
```
Comment (mimaison) [inline docs/ops.html]: This line is not correct and should be removed. You don't provide the path when defining the provider, you only provide it when using the provider to retrieve a value.

**Review by mimaison — COMMENTED** [review]
Thanks for the updates. I left a few comments on the content itself.

I also wonder if we should move this section at the end of Configuration instead of having it in the middle of the Operations section. WDYT?

**Comment by PaulRMellor** [comment]
Thanks for the additional comments @mimaison 
The content now sits in the config section.

File: docs/toc.html [hunk 16]
```diff
@@ -50,13 +50,23 @@
...
+                
                 <li><a href="#streamsconfigs">3.6 Kafka Streams Configs</a>
                 <li><a href="#adminclientconfigs">3.7 AdminClient Configs</a>
                 <li><a href="#mirrormakerconfigs">3.8 MirrorMaker Configs</a>
                 <li><a href="#systemproperties">3.9 System Properties</a>
-                <li><a href="#tieredstorageconfigs">3.10 Tiered Storage Configs</a></li>
+                <li><a href="#tieredstorageconfigs">3.10 Tiered Storage Configs</a>
+                <li><a href="#config_providers">3.11 Configuration Providers</a>
+                    <ul>
+                        <li><a href="#using_providers">Using</a>
```
Comment (mimaison) [inline docs/toc.html]: Should this be `Using Configuration Providers` to match the section title?

File: docs/ops.html [hunk 17]
```diff
@@ -1214,9 +1214,9 @@ <h4 class="anchor-heading"><a id="multitenancy-more" class="anchor-link"></a><a
   </p>
 
 
-  <h3 class="anchor-heading"><a id="config" class="anchor-link"></a><a href="#config">6.5 Kafka Configuration</a></h3>
```
Comment (mimaison) [inline docs/ops.html]: Should we undo the changes in this file as the new section is in `configuration.html`? If so let's undo the changes in `toc.html` too.

File: docs/configuration.html [hunk 18]
```diff
@@ -308,6 +308,157 @@ <h3 class="anchor-heading"><a id="tieredstorageconfigs" class="anchor-link"></a>
...
+  </p>
+  <p>
+    Pass additional parameters to each provider as <code>config.providers.&lt;provider_alias&gt;.param.&lt;name&gt;=&lt;value&gt;</code>:
+  </p>
+  <pre><code>config.providers.provider1.param.param1=value1
+config.providers.provider1.param.param2=value2</code></pre>
+  
+  <h4 class="anchor-heading">
+    <a id="config_provider" class="anchor-link"></a>
+    <a href="#config_provider">ConfigProvider</a>
```
Comment (mimaison) [inline docs/configuration.html]: This section only repeats things already mentioned above. Do we really need it?

File: docs/configuration.html [hunk 19]
```diff
@@ -308,6 +308,157 @@ <h3 class="anchor-heading"><a id="tieredstorageconfigs" class="anchor-link"></a>
...
+  <p>The <code>FileConfigProvider</code> retrieves configuration data from a single properties file.</p>
+  <p>This provider is useful for loading configuration data from mounted files.</p>
+  
+  <p><b>Example <code>FileConfigProvider</code> configuration</b></p>
+  <pre><code>config.providers=fileProvider
+config.providers.fileProvider.class=org.apache.kafka.common.config.provider.FileConfigProvider</code></pre>
+  
+  <h4 class="anchor-heading">
+    <a id="ref_config_provider" class="anchor-link"></a>
+    <a href="#ref_config_provider">Referencing configuration providers</a>
```
Comment (mimaison) [inline docs/configuration.html]: I wonder if it would read better if the syntax to use each provider were in their respective sections? WDYT?

File: docs/configuration.html [hunk 20]
```diff
@@ -308,6 +308,157 @@ <h3 class="anchor-heading"><a id="tieredstorageconfigs" class="anchor-link"></a>
...
+    The <code>DirectoryConfigProvider</code> retrieves configuration data from files stored in a specified directory.
+  </p>
+  <p>
+    Each file represents a key, and its content is the value. You specify the path to the directory where the configuration files are stored using the <code>dir</code> parameter. This provider is useful for loading multiple configuration files and for organizing configuration data into separate files.
+  </p>
+  
+  <p><b>Example <code>DirectoryConfigProvider</code> configuration</b></p>
+  <pre><code>config.providers=dirProvider
+config.providers.dirProvider.class=org.apache.kafka.common.config.provider.DirectoryConfigProvider
+config.providers.dirProvider.param.dir=/path/to/config/directory</code></pre>
```
Comment (mimaison) [inline docs/configuration.html]: This is not correct. The path is only provided when using the provider.

File: docs/configuration.html [hunk 21]
```diff
@@ -308,6 +308,157 @@ <h3 class="anchor-heading"><a id="tieredstorageconfigs" class="anchor-link"></a>
...
+  
+  <p><b>Example custom provider configuration</b></p>
+  <pre><code>config.providers=customProvider
+config.providers.customProvider.class=com.example.customProvider
+config.providers.customProvider.param.param1=value1
+config.providers.customProvider.param.param2=value2</code></pre>
+  
+  <h4 class="anchor-heading">
+    <a id="directory_config_provider" class="anchor-link"></a>
+    <a href="#directory_config_provider">DirectoryConfigProvider</a>
```
Comment (mimaison) [inline docs/configuration.html:361]: This provider has a configurable parameter. I think it would be nice to mention it here:
https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/config/provider/DirectoryConfigProvider.java#L48-L50

File: docs/configuration.html [hunk 22]
```diff
@@ -308,6 +308,157 @@ <h3 class="anchor-heading"><a id="tieredstorageconfigs" class="anchor-link"></a>
...
+  </p>
+  
+  <p><b>Example <code>DirectoryConfigProvider</code> configuration</b></p>
+  <pre><code>config.providers=dirProvider
+config.providers.dirProvider.class=org.apache.kafka.common.config.provider.DirectoryConfigProvider
+config.providers.dirProvider.param.dir=/path/to/config/directory</code></pre>
+  
+  <h4 class="anchor-heading">
+    <a id="env_var_config_provider" class="anchor-link"></a>
+    <a href="#env_var_config_provider">EnvVarConfigProvider</a>
```
Comment (mimaison) [inline docs/configuration.html:378]: This provider has a configurable parameter. I think it would be nice to mention it here:
https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/config/provider/EnvVarConfigProvider.java#L42-L44

File: docs/configuration.html [hunk 23]
```diff
@@ -308,6 +308,157 @@ <h3 class="anchor-heading"><a id="tieredstorageconfigs" class="anchor-link"></a>
...
+    This provider is useful for configuring applications running in containers, for example, to load certificates or JAAS configuration from environment variables mapped from secrets.
+  </p>
+  
+  <p><b>Example <code>EnvVarConfigProvider</code> configuration</b></p>
+  <pre><code>config.providers=envVarProvider
+config.providers.envVarProvider.class=org.apache.kafka.common.config.provider.EnvVarConfigProvider</code></pre>
+  
+  <h4 class="anchor-heading">
+    <a id="file_config_provider" class="anchor-link"></a>
+    <a href="#file_config_provider">FileConfigProvider</a>
```
Comment (mimaison) [inline docs/configuration.html:395]: This provider has a configurable parameter. I think it would be nice to mention it here:
https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/config/provider/FileConfigProvider.java#L45-L47

File: docs/configuration.html [hunk 24]
```diff
@@ -308,6 +308,157 @@ <h3 class="anchor-heading"><a id="tieredstorageconfigs" class="anchor-link"></a>
...
+  
+  <p><b>Example Kafka Connect configuration with a <code>FileConfigProvider</code></b></p>
+  <pre><code>config.providers=fileProvider
+config.providers.fileProvider.class=org.apache.kafka.common.config.provider.FileConfigProvider</code></pre>
+  
+  <p>Next, reference the properties from the file in the connector configuration.</p>
+  <p><b>Example connector configuration referencing file properties</b></p>
+  <pre><code>database.user=${fileProvider:/path/to/connector-credentials.properties:dbUsername}
+database.password=${fileProvider:/path/to/connector-credentials.properties:dbPassword}</code></pre>
+  <p>The configuration provider reads and extracts the values from the properties file.</p>
```
Comment (mimaison) [inline docs/configuration.html]: I wonder if we add `At runtime,` at the beginning of the sentence.

**Review by mimaison — COMMENTED** [review]
Thanks for the updates. I left a few more suggestions

File: docs/toc.html (hunk 16 above)
Comment (PaulRMellor) [reply docs/toc.html]: I think it's fine as it is. I tried a full title with this and referencing and they looked too long in the toc and the concept gets a bit lost. 

image

with full names:

image

Comment (mimaison) [reply docs/toc.html]: I'm not a native English speaker and just "Using" looks weird to me. If we don't want the full title because it's too long, would "Usage" work? If you think "Using" is fine, then I'll trust you.

Comment (PaulRMellor) [reply docs/toc.html]: Usage is okay but a bit formal and less direct than Using. Using is more action-oriented and fits with referencing (gerund) later. Most technical documents use gerunds for headings. My vote is Using.

Comment (mimaison) [reply docs/toc.html]: Ok, thanks for the explanation

File: docs/configuration.html (hunk 19 above)
Comment (PaulRMellor) [reply docs/configuration.html]: I think the split make sense. We're doing this somewhere else. Also, people just looking to see how they reference the external properties can come straight here from the TOC.

**Comment by PaulRMellor** [comment]
Thanks @mimaison -- we mention the parameters now

File: docs/configuration.html (hunk 18 above)
Comment (mimaison) [reply docs/configuration.html]: Let's also remove the header in `toc.html`

File: docs/configuration.html (hunk 19 above)
Comment (mimaison) [reply docs/configuration.html]: My thinking is that in most cases (configuring clients, brokers, Streams applications, the Connect runtime), it's the same person doing both actions and doing them at the same time. The only outlier scenario is using configuration providers in a connector configuration as the provider first needs to be defined in the runtime and used at a later time when creating a connector.

**Comment by PaulRMellor** [comment]
Thanks @mimaison -- I've now moved the syntax to sit with provider. I renamed the final section "Example: Referencing Files" and because of this renamed "Using" to "Using Configuration Providers" (I didn't like all the _Configuration Providers_ in the headings when we considered this before. I thought it got in the way.)

**Review by mimaison — APPROVED** [review]
LGTM, thanks for the PR
